virttest/utils_time.py

```diff
@@ -0,0 +1,259 @@
+    :param timeout: Timeout to execute command
+    :param session: ShellSession or None
+
+    :return: Command output string
+    """
+    if session:
+        ret = session.cmd_output(cmd, timeout=timeout)
+    else:
+        ret = decode_to_text(process.system_output(cmd))
+    target = session and "guest" or "host"
+    logging.debug("(%s) Execute command('%s')" % (target, cmd))
+    return ret
+
+
+@error_context.context_aware
+def verify_clocksource(expected, session=None):
+    """
+    Verify if host/guest use the expected clocksource
+    :param expected: Expected clocksource
+    :param session: VM session
+    """
+    error_context.context("Check the current clocksource", logging.info)
+    cmd = "cat /sys/devices/system/clocksource/"
+    cmd += "clocksource0/current_clocksource"
+    if expected not in execute(cmd, session=session):
+        return False
+    return True
+
+
+@error_context.context_aware
+def sync_time_with_ntp(session=None):
+    """
+    Sync guest or host time with ntp server
+    :param session: VM session or None
+    """
+    error_context.context("Sync time from ntp server", logging.info)
+    cmd = "ntpdate clock.redhat.com; hwclock -w"
+    return execute(cmd, session)
+
+
+@error_context.context_aware
+def update_clksrc(vm, image_filename, grub_file, kernel_cfg_pattern, clksrc=None):
+    """
+    Update linux guest's clocksource and re-boot guest
+
+    :params vm: Virtual machine for vm
+    :params image_filename: Current image's file path
+    :params grub_file: VM's grub file path
+    :params kernel_cfg_pattern: Pattern to get current kernel line
+    :params clksrc: Expected clocksource, 'kvm-clock' by default
+    """
+    if vm.is_alive():
+        error_context.context("Shutdown guest", logging.info)
+        vm.destroy()
+    _update_clksrc_with_kernel_line(image_filename, grub_file,
+                                    kernel_cfg_pattern, clksrc)
+    params = vm.get_params()
+    if 'fedora' in params["os_variant"] and clksrc:
+        cpu_model_flags = params.get["cpu_model_flags"]
+        params["cpu_model_flags"] = cpu_model_flags + "-kvmclock"
+    error_context.context("Boot the guest", logging.info)
+    vm.create(params=params)
+    vm.verify_alive()
+
+
+@error_context.context_aware
+def _update_clksrc_with_kernel_line(image_filename, grub_file,
+                                    kernel_cfg_pattern, clksrc=None):
+    """
+    Update clocksource in kernel line with libguestfs
+
+    :params image_filename: Current image's file path
+    :params grub_file: VM's grub file path
+    :params kernel_cfg_pattern: Pattern to get current kernel line
+    :param clksrc: Expected clocksource. Set clocksource=None when finish
+                   test, clocksource should be 'kvm-clock' by default
+    """
+    error_context.context("Update guest kernel cli to '%s'" % (clksrc or
+                          "kvm-clock"), logging.info)
+    disk_obj = utils_disk.GuestFSModiDisk(image_filename)
```

Comment:
is it mandatory to use libguestfs ? can we modify kernel commandline using API `update_boot_option` in utils_misc.py,